README.md
<!--
 * @Author: meetqy
 * @since: 2019-11-05 17:45:28
 * @lastTime: 2019-11-06 09:50:48
 * @LastEditors: meetqy
 -->
# flutter_collapse

flutter_collapse折叠面板

## 目录

- [预览](#预览)
- [使用](#使用)
- [参数说明](#参数说明)
- [欢迎PR](#欢迎PR)
- [使用许可](#使用许可)

## 预览

![](preview.gif)

## 使用说明

```dart
Collapse(
  title: Container(
    child: Text('折叠面板 - 默认')
  ),
  body: Text('这就是一个demo'),
  value: true,
),
```

## 参数说明

参数 | 说明
---- | ----
title\<Widget> | 标题,
icon\<Widget> | 右侧图标。默认值：Icon(Icons.expand_more, size: 24,)
value\<bool> | 是否展开。 默认值：true
body\<Widget> | 内容
onChange\<Function(bool)> | 状态改变 返回当前value
showBorder\<bool> | 是否显示border 默认：true
padding\<EdgeInsets> | 内边距 默认值：EdgeInsets.symmetric(vertical: 10, horizontal: 15),

## 维护者

[@meetqy](https://github.com/meetqy).

## 欢迎PR

非常欢迎你的加入! [提一个Issue](https://github.com/meetqy/flutter_collapse/issues/new) 或者提交一个 Pull Request.

## 使用许可

[MIT](LICENSE) © MEETQY


CHANGELOG.md
<!--
 * @Author: meetqy
 * @since: 2019-11-05 17:45:28
 * @lastTime: 2019-11-06 11:40:11
 * @LastEditors: meetqy
 -->
## [0.1.1] - 优化注释

* 优化注释



## [0.1.0] - 修改example

* 修改example
* 修改desc

## [0.0.1] - first push

* first push


lib/animate_rotate.dart
/*
 * @Author: meetqy
 * @since: 2019-10-22 11:24:46
 * @lastTime: 2019-11-06 11:39:10
 * @LastEditors: meetqy
 */
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

class AnimatedRotate extends ImplicitlyAnimatedWidget {
  /// 自定义旋转动画
  /// 参考：自带动画 AnimatedOpacity
  /// 详细注解查看  AnimatedOpacity
  /// ``` dart
  /// @params rotate 计算方式：x * math.pi * 2
  /// ```
  const AnimatedRotate({
    Key key,
    this.child,
    @required this.rotate,
    Curve curve = Curves.linear,
    @required Duration duration,
  })  : assert(rotate != null),
        super(key: key, curve: curve, duration: duration);

  final Widget child;

  final double rotate;

  @override
  _AnimatedRotateState createState() => _AnimatedRotateState();

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(DoubleProperty('rotate', rotate));
  }
}

class _AnimatedRotateState
    extends ImplicitlyAnimatedWidgetState<AnimatedRotate> {
  Tween<double> _rotate;
  Animation<double> _rotateAnimation;

  @override
  void forEachTween(TweenVisitor<dynamic> visitor) {
    _rotate = visitor(
        _rotate, widget.rotate, (dynamic value) => Tween<double>(begin: value));
  }

  @override
  void didUpdateTweens() {
    _rotateAnimation = animation.drive(_rotate);
  }

  @override
  Widget build(BuildContext context) {
    return RotationTransition(
      turns: _rotateAnimation,
      alignment: Alignment.center,
      child: widget.child,
    );
  }
}


lib/flutter_collapse.dart
library flutter_collapse;

import 'package:flutter/material.dart';
import 'package:flutter_collapse/animate_rotate.dart';

class Collapse extends StatelessWidget {
  /// 折叠面板
  const Collapse({
    Key key,
    this.title,
    this.icon = const Icon(
      Icons.expand_more,
      size: 24,
    ),
    this.value = true,
    this.body,
    this.onChange,
    this.showBorder = true,
    this.padding = const EdgeInsets.symmetric(vertical: 10, horizontal: 15),
  }) : super(key: key);

  /// 标题
  final Widget title;

  /// 是否展开 true:展开 false:收起 默认值:true
  final bool value;

  /// 内容
  final Widget body;

  /// 回调 返回当前value
  final Function(bool) onChange;

  /// icon 默认值：Icon(Icons.expand_more, size: 24,),
  final Widget icon;

  /// padding 默认值：EdgeInsets.symmetric(vertical: 10, horizontal: 15)
  final EdgeInsets padding;

  /// 是否显示border
  final bool showBorder;

  static const Color borderColor = Color.fromRGBO(235, 237, 240, 1);
  static const Duration sleep = const Duration(milliseconds: 120);

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.white,
      key: key,
      child: Column(
        children: <Widget>[
          /// title
          InkWell(
            child: Container(
              padding: const EdgeInsets.symmetric(vertical: 10, horizontal: 15),
              decoration: BoxDecoration(
                  border: Border(
                      bottom: BorderSide(
                          color:
                              showBorder ? borderColor : Colors.transparent))),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: <Widget>[
                  Container(
                    child: title,
                  ),
                  AnimatedRotate(
                    duration: sleep,
                    child: icon,
                    rotate: value ? 0.5 : 1,
                  )
                ],
              ),
            ),
            onTap: () => onChange == null ? () {} : onChange(!value),
          ),

          _Body(
            child: Container(padding: padding, child: body),
            sleep: sleep,
            value: value,
          )
        ],
      ),
    );
  }
}

/// 内容区域
/// 为了获取当前内容的高度 实现展开收起动画，所以拆成状态组件
class _Body extends StatefulWidget {
  _Body({
    this.child,
    this.value,
    @required this.sleep,
    Key key,
  }) : super(key: key);

  final Widget child;
  final bool value;
  final Duration sleep;

  @override
  _BodyState createState() => _BodyState();
}

class _BodyState extends State<_Body> with TickerProviderStateMixin {
  Widget build(BuildContext context) {
    return Container(
      child: AnimatedSize(
          vsync: this,
          duration: widget.sleep,
          child: Container(
            constraints: BoxConstraints(
              maxHeight: !widget.value ? 0 : double.infinity,
            ),
            child: Row(
              children: <Widget>[
                Expanded(
                  child: widget.child,
                )
              ],
            ),
          )),
    );
  }
}


example/index.dart
/*
 * @Author: meetqy
 * @since: 2019-10-22 09:59:58
 * @lastTime: 2019-11-05 17:51:08
 * @LastEditors: meetqy
 */
import 'package:flutter/material.dart';
import 'package:flutter_collapse/flutter_collapse.dart';

class ExampleCollapse extends StatefulWidget {
  ExampleCollapse({Key key}) : super(key: key);

  @override
  _ExampleCollapseState createState() => _ExampleCollapseState();
}

class _ExampleCollapseState extends State<ExampleCollapse> {
  bool status1 = true;
  bool status2 = true;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Color.fromRGBO(245, 245, 245, 1),
      appBar: AppBar(
        title: Text('collapse'),
      ),
      body: Container(
        margin: EdgeInsets.only(top: 20),
        child: SingleChildScrollView(
          child: Column(
            children: <Widget>[
              Collapse(
                title: Container(
                  child: Text('折叠面板 - 默认')
                ),
                body: Text('这就是一个demo'),
                value: status1,
                onChange: (bool value) {
                  setState(() {
                    status1 = value;
                  });
                },
              ),

              Container(
                margin: EdgeInsets.only(top: 20),
                child: Collapse(
                  padding: EdgeInsets.all(0),
                  title: Text('折叠面板 - 自定义'),
                  body: Container(
                    color: Colors.blue,
                    padding: EdgeInsets.symmetric(vertical: 10, horizontal: 15),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: <Widget>[
                        Text('这就是一个demo这就是一个demo这就是一个demo这就是一个demo这就是一个demo'),
                        Text('这就是一个demo这就是一个demo这就是一个demo'),
                        Text('这就是一个demo这就是一个demo'),
                        Row(
                          mainAxisAlignment: MainAxisAlignment.spaceBetween,
                          children: <Widget>[
                            CircularProgressIndicator(
                              backgroundColor: Colors.yellow,
                            ),
                            CircularProgressIndicator(
                              backgroundColor: Colors.green,
                            ),
                            CircularProgressIndicator(
                              backgroundColor: Colors.red,
                            )
                          ],
                        ),
                      ],
                    ),
                  ),
                  value: status2,
                  onChange: (bool value) {
                    setState(() {
                      status2 = value;
                    });
                  },
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
